### PR TITLE
FFWEB-1612: Check FTP connection

### PR DESCRIPTION
**Pull request by a-laurowski**
- Description: 
Added Check FTP connection
- Tested with Magento editions/versions: 
2.3.0
- Tested with PHP versions: 
7.2

### CODE REVIEW

File: src/Controller/Adminhtml/TestConnection/TestFtpConnection.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    /** @var  FtpUploader */
+    private $ftpUploader;
+
+    /** @var FtpConfig */
+    private $ftpConfig;
+
+    public function __construct(
+        Action\Context $context,
+        JsonFactory $jsonResultFactory,
+        FtpUploader $ftpUploader,
```
Comment (fmarangi) [inline src/Controller/Adminhtml/TestConnection/TestFtpConnection.php]: Not used

File: src/Controller/Adminhtml/TestConnection/TestFtpConnection.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        return $this->jsonResultFactory->create()->setData(['message' => $message]);
+    }
+
+    private function getConfig(array $params): array
+    {
+        $prefix   = 'ff_upload_';
+        $filtered = array_filter($params, function (string $key) use ($prefix): bool {
+            return (bool) preg_match("/$prefix/", $key);
```
Comment (fmarangi) [inline src/Controller/Adminhtml/TestConnection/TestFtpConnection.php]: Let's add a `^` for safety => `"#^{$prefix}#"`

File: src/Model/FtpUploader.php [hunk 3]
```diff
@@ -22,6 +22,26 @@ public function __construct(FtpConfig $config, IoInterface $client)
...
+     */
+    public function testConnection(array $params): void
+    {
+        $fileName = 'testconnection';
+        $this->client->open($params);
+        //check write permission
+        $this->client->write($fileName, '');
+        $this->client->rm($fileName);
+
+        $this->client->close();
```
Comment (fmarangi) [inline src/Model/FtpUploader.php]: Can we pack the last 3 lines in a try/finally block? Otherwise the connection stays open.

File: README.md [hunk 4]
```diff
@@ -128,7 +128,8 @@ This option configures the connection with the FACT-Finder system via FTP. Shop
 FACT-Finder using FTP. FACT-Finder needs to be up to date on the product data, to ensure that components like the search work as intended.
 
 Enter an FTP-server to which the CSV file is uploaded automatically. The URL needs to be entered without the protocol
-prefix (`ftp://`) and without the slash at the end.
+prefix (`ftp://`) and without the slash at the end. If you are not sure if Magento will be able to connect to your FTP server, please use
```
Comment (fmarangi) [inline README.md]: About this part ... what do you think about stripping the slash / protocoll from the string? It's probably less confusion for the users.

File: src/Model/Config/FtpConfig.php [hunk 5]
```diff
@@ -24,9 +24,9 @@ public function toArray(): array
             'host'     => $this->getConfig('host'),
             'user'     => $this->getConfig('user'),
             'password' => $this->getConfig('password'),
-            'ssl'      => true,
+            'ssl'      => $this->getConfig('use_ssl'),
             'passive'  => true,
-            'port'     => 21,
+            'port'     => $this->getConfig('port') ?? 21,
```
[the remark this answers is not in the document]
Comment (a-laurowski) [reply src/Model/Config/FtpConfig.php]: changed

File: src/Controller/Adminhtml/TestConnection/TestFtpConnection.php (hunk 2 above)
Comment (a-laurowski) [reply src/Controller/Adminhtml/TestConnection/TestFtpConnection.php]: changed

File: src/Model/FtpUploader.php (hunk 3 above)
Comment (a-laurowski) [reply src/Model/FtpUploader.php]: changed

File: README.md (hunk 4 above)
Comment (a-laurowski) [reply README.md]: added `trimProtocol` function in `FtpUploader`, updated README.md

**Pull request merged**